File: clients/android/app/src/main/java/local/visual/corepose/VideoPose.kt
package local.visual.corepose

import android.graphics.Bitmap
import android.graphics.Matrix
import android.media.MediaMetadataRetriever
import android.os.Build
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.vision.poselandmarker.PoseLandmarker

data class PoseLandmarkers(
    val video: PoseLandmarker,
    val crop: PoseLandmarker,
) {
    fun close() {
        video.close()
        crop.close()
    }
}

object VideoPose {
    const val MAX_MS = 120_000L
    /** Match Windows normalize long-edge cap for MediaPipe cost. */
    const val ANALYZE_MAX_EDGE = 720
    /** Cap sample rate even when source is 60fps (Windows ingest is 30fps). */
    const val ANALYZE_MAX_FPS = 30.0

    fun analyze(
        retriever: MediaMetadataRetriever,
        markers: PoseLandmarkers,
        device: String,
        startMs: Long = 0L,
        endMs: Long = MAX_MS,
        seeds: List<SeedMark> = emptyList(),
        seedBox: NormBox? = null,
        fpsIn: Double = 30.0,
        onProgress: (doneMs: Long, totalMs: Long) -> Unit,
    ): List<PoseFrame> {
        val durationMs = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
            ?.toLongOrNull()
            ?.coerceAtLeast(0L)
            ?: 0L
        val srcFps = if (fpsIn > 1.0) fpsIn else 30.0
        val sampleFps = minOf(srcFps, ANALYZE_MAX_FPS)
        val lo = startMs.coerceAtLeast(0L)
        val hi = endMs.coerceAtMost(durationMs.coerceAtMost(lo + MAX_MS)).coerceAtLeast(lo)
        val rotation = if (Build.VERSION.SDK_INT >= 27) {
            0
        } else {
            retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_VIDEO_ROTATION)
                ?.toIntOrNull()
                ?: 0
        }
        val timedSeeds = ClipRange.collectSeeds(seeds, seedBox)
        val halfMs = ClipRange.halfMs(sampleFps)
        val stepMs = ClipRange.strideMs(sampleFps)
        val locator = PoseLocator(device)
        val frames = ArrayList<PoseFrame>()
        var tMs = lo
        while (tMs <= hi) {
            onProgress(tMs - lo, (hi - lo).coerceAtLeast(1L))
            val option = MediaMetadataRetriever.OPTION_CLOSEST
            val raw = retriever.getFrameAtTime(tMs * 1000, option)
            if (raw != null) {
                val bmp = raw.rotated(rotation)
                if (bmp !== raw) {
                    raw.recycle()
                }
                val (work, scaled) = bmp.forAnalyze(ANALYZE_MAX_EDGE)
                val inferred = locator.inferFrame(work, tMs.toDouble(), timedSeeds, seedBox, halfMs, markers)
                frames.add(inferred.remapToSize(bmp.width, bmp.height))
                if (scaled) {
                    work.recycle()
                }
                bmp.recycle()
            }
            tMs += stepMs
        }
        return PoseTrack.fillLowScore(PoseTrack.stabilizeWeakJoints(frames))
    }

    fun analyzeBitmap(
        bitmap: Bitmap,
        markers: PoseLandmarkers,
        device: String,
        seeds: List<SeedMark> = emptyList(),
        seedBox: NormBox? = null,
    ): List<PoseFrame> {
        val locator = PoseLocator(device)
        val timed = ClipRange.collectSeeds(seeds, seedBox)
        val (work, scaled) = bitmap.forAnalyze(ANALYZE_MAX_EDGE)
        val inferred = locator.inferFrame(work, 0.0, timed, seedBox, 1e9, markers)
        val frame = inferred.remapToSize(bitmap.width, bitmap.height)
        if (scaled) {
            work.recycle()
        }
        return PoseTrack.fillLowScore(listOf(frame))
    }
}

class PoseLocator(private val device: String) {
    private var hist: FloatArray? = null
    private var box: PixelBox? = null
    private var lastMapped: MappedPeople? = null
    private var lastTs = -1L

    fun inferFrame(
        bmp: Bitmap,
        tMs: Double,
        seeds: List<SeedMark>,
        seedBox: NormBox?,
        halfMs: Double,
        markers: PoseLandmarkers,
    ): PoseFrame {
        val w = bmp.width
        val h = bmp.height
        val pixels = IntArray(w * h)
        bmp.getPixels(pixels, 0, w, 0, 0, w, h)
        val hit = ClipRange.seedAt(tMs, seeds, halfMs)
        when {
            hit != null -> {
                box = PersonRoi.denormBox(hit.box, w, h)
                hist = PersonRoi.buildHist(pixels, w, h, box!!)
            }
            hist == null && seedBox != null -> {
                box = PersonRoi.denormBox(seedBox, w, h)
                hist = PersonRoi.buildHist(pixels, w, h, box!!)
            }
            hist != null && box != null -> {
                box = PersonRoi.search(pixels, w, h, hist!!, box!!)
            }
        }
        val mapped = detect(bmp, tMs.toLong(), box, markers)
        if (mapped.poses.isNotEmpty() && mapped.bboxes.isNotEmpty()) {
            val det = mapped.bboxes[0]
            box = PixelBox(det.x1.toDouble(), det.y1.toDouble(), det.x2.toDouble(), det.y2.toDouble())
            val nextHist = PersonRoi.buildHist(pixels, w, h, box!!)
            hist = hist?.let { PersonRoi.blendHist(it, nextHist) } ?: nextHist
            lastMapped = mapped
        } else {
            lastMapped = null
        }
        val pose = mapped.poses.firstOrNull()
        val blaze = mapped.blazePeople.firstOrNull().orEmpty()
        return PoseFrame(
            tMs.toLong(),
            if (pose != null) listOf(pose) else emptyList(),
            w,
            h,
            blaze,
            mapped.bboxes.firstOrNull(),
        )
    }

    private fun detect(
        bmp: Bitmap,
        tMs: Long,
        roi: PixelBox?,
        markers: PoseLandmarkers,
    ): MappedPeople {
        var mapped: MappedPeople? = null
        if (roi != null) {
            val cropped = detectCrop(bmp, roi, markers.crop)
            val ok = PoseFilter.isPlausible(cropped, bmp.width, bmp.height) || cropped.poses.isNotEmpty()
            if (ok && !PoseFilter.cropDisagreesWithPrev(lastMapped, cropped)) {
                mapped = cropped
            }
        }
        if (mapped == null) {
            mapped = detectFull(bmp, tMs, markers.video)
            mapped = PoseFilter.pickPrimary(mapped, lastMapped?.bboxes?.firstOrNull())
            if (!PoseFilter.isPlausible(mapped, bmp.width, bmp.height)) {
                val retryBox = mapped.bboxes.firstOrNull() ?: lastMapped?.bboxes?.firstOrNull()
                if (retryBox != null) {
                    val retry = detectCrop(bmp, PoseFilter.cropBoxForRetry(retryBox, bmp.width, bmp.height), markers.crop)
                    val retryOk = PoseFilter.isPlausible(retry, bmp.width, bmp.height) || retry.poses.isNotEmpty()
                    if (retryOk && !PoseFilter.cropDisagreesWithPrev(lastMapped, retry)) {
                        mapped = retry
                    }
                }
                if (!PoseFilter.isPlausible(mapped, bmp.width, bmp.height) && mapped.poses.isEmpty()) {
                    mapped = MappedPeople(emptyList(), emptyList(), emptyList())
                }
            }
        }
        return PoseFilter.emaSmooth(lastMapped, mapped)
    }

    private fun detectCrop(src: Bitmap, roi: PixelBox, marker: PoseLandmarker): MappedPeople {
        val spec = PersonRoi.cropRect(roi, src.width, src.height)
        val cw = (spec.x2 - spec.ox).coerceAtLeast(2)
        val ch = (spec.y2 - spec.oy).coerceAtLeast(2)
        var crop = if (cw >= src.width - 1 && ch >= src.height - 1) {
            src
        } else {
            Bitmap.createBitmap(src, spec.ox, spec.oy, cw, ch)
        }
        if (spec.scale > 1.0 + 1e-6) {
            val sw = (cw * spec.scale).toInt().coerceAtLeast(2)
            val sh = (ch * spec.scale).toInt().coerceAtLeast(2)
            val next = Bitmap.createScaledBitmap(crop, sw, sh, true)
            if (crop !== src) {
                crop.recycle()
            }
            crop = next
        }
        val result = marker.detect(BitmapImageBuilder(crop).build())
        val mapped = PoseMapper.fromLandmarks(result.landmarks(), crop.width, crop.height, device)
        if (crop !== src) {
            crop.recycle()
        }
        val picked = PoseFilter.pickPrimary(mapped, lastMapped?.bboxes?.firstOrNull())
        return remap(picked, spec)
    }

    private fun detectFull(src: Bitmap, tMs: Long, marker: PoseLandmarker): MappedPeople {
        val ts = maxOf(tMs, lastTs + 1)
        lastTs = ts
        val result = marker.detectForVideo(BitmapImageBuilder(src).build(), ts)
        return PoseMapper.fromLandmarks(result.landmarks(), src.width, src.height, device)
    }

    private fun remap(mapped: MappedPeople, spec: CropSpec): MappedPeople {
        fun pt(x: Float, y: Float): Pair<Float, Float> {
            val (nx, ny) = PersonRoi.remapPoint(x.toDouble(), y.toDouble(), spec.ox, spec.oy, spec.scale)
            return nx.toFloat() to ny.toFloat()
        }
        val poses = mapped.poses.map { pose ->
            CocoPose(
                pose.keypoints.map { kp ->
                    val (x, y) = pt(kp.x, kp.y)
                    CocoKeypoint(x, y, kp.confidence)
                },
            )
        }
        val blaze = mapped.blazePeople.map { person ->
            person.map { joint ->
                val (x, y) = pt(joint.x, joint.y)
                BlazeJoint(x, y, joint.z, joint.confidence)
            }
        }
        val boxes = mapped.bboxes.map { box ->
            val (x1, y1) = pt(box.x1, box.y1)
            val (x2, y2) = pt(box.x2, box.y2)
            BBox(x1, y1, x2, y2)
        }
        return MappedPeople(poses, blaze, boxes)
    }
}

private fun Bitmap.rotated(degrees: Int): Bitmap {
    if (degrees % 360 == 0) {
        return this
    }
    val matrix = Matrix().apply { postRotate(degrees.toFloat()) }
    return Bitmap.createBitmap(this, 0, 0, width, height, matrix, true)
}

/** Downscale so the long edge is at most [maxEdge]. Second value true when a new bitmap was created. */
private fun Bitmap.forAnalyze(maxEdge: Int): Pair<Bitmap, Boolean> {
    val longEdge = maxOf(width, height)
    if (longEdge <= maxEdge) {
        return this to false
    }
    val scale = maxEdge.toFloat() / longEdge.toFloat()
    val w = (width * scale).toInt().coerceAtLeast(2)
    val h = (height * scale).toInt().coerceAtLeast(2)
    return Bitmap.createScaledBitmap(this, w, h, true) to true
}

private fun PoseFrame.remapToSize(origW: Int, origH: Int): PoseFrame {
    if (width == origW && height == origH) {
        return this
    }
    val sx = origW.toFloat() / width.toFloat()
    val sy = origH.toFloat() / height.toFloat()
    fun scalePose(pose: CocoPose): CocoPose {
        return CocoPose(
            pose.keypoints.map { kp ->
                CocoKeypoint(kp.x * sx, kp.y * sy, kp.confidence)
            },
        )
    }
    fun scaleBlaze(joints: List<BlazeJoint>): List<BlazeJoint> {
        return joints.map { j -> BlazeJoint(j.x * sx, j.y * sy, j.z, j.confidence) }
    }
    return PoseFrame(
        tMs = tMs,
        poses = poses.map(::scalePose),
        width = origW,
        height = origH,
        blaze33 = scaleBlaze(blaze33),
        bbox = bbox?.let { BBox(it.x1 * sx, it.y1 * sy, it.x2 * sx, it.y2 * sy) },
    )
}

File: clients/ios/VisualPose/VideoPose.swift
import AVFoundation
import Foundation
import UIKit

enum VideoPose {
    static let maxMs: Int64 = 120_000

    static func analyze(
        asset: AVAsset,
        markers: PoseLandmarkers,
        device: String,
        startMs: Int64 = 0,
        endMs: Int64 = maxMs,
        seeds: [SeedMark] = [],
        seedBox: NormBox? = nil,
        fpsIn: Double = 30,
        onProgress: (Int64, Int64) -> Void
    ) -> [PoseFrame] {
        let durationSec = CMTimeGetSeconds(asset.duration)
        let durationMs = durationSec.isFinite ? max(Int64(durationSec * 1000), 0) : 0
        let fps = fpsIn > 1.0 ? fpsIn : 30.0
        let lo = max(startMs, 0)
        let hi = max(min(endMs, min(durationMs, lo + maxMs)), lo)
        let timedSeeds = ClipRange.collectSeeds(seeds, seedBox: seedBox)
        let halfMs = ClipRange.halfMs(fps: fps)
        let stepMs = ClipRange.strideMs(fps: fps)
        let locator = PoseLocator(device: device)
        let generator = AVAssetImageGenerator(asset: asset)
        generator.appliesPreferredTrackTransform = true
        generator.requestedTimeToleranceBefore = .zero
        generator.requestedTimeToleranceAfter = .zero
        var frames: [PoseFrame] = []
        var tMs = lo
        while tMs <= hi {
            onProgress(tMs - lo, max(hi - lo, 1))
            if let image = frameImage(generator, tMs: tMs) {
                frames.append(
                    locator.inferFrame(image, tMs: Double(tMs), seeds: timedSeeds, seedBox: seedBox, halfMs: halfMs, markers: markers)
                )
            }
            tMs += stepMs
        }
        return PoseTrack.fillLowScore(PoseTrack.stabilizeWeakJoints(frames))
    }

    static func analyzeBitmap(
        _ image: UIImage,
        markers: PoseLandmarkers,
        device: String,
        seeds: [SeedMark] = [],
        seedBox: NormBox? = nil
    ) -> [PoseFrame] {
        let locator = PoseLocator(device: device)
        let timed = ClipRange.collectSeeds(seeds, seedBox: seedBox)
        return PoseTrack.fillLowScore(
            PoseTrack.stabilizeWeakJoints(
                [locator.inferFrame(image, tMs: 0, seeds: timed, seedBox: seedBox, halfMs: 1e9, markers: markers)]
            )
        )
    }

    private static func frameImage(_ generator: AVAssetImageGenerator, tMs: Int64) -> UIImage? {
        let time = CMTime(value: tMs, timescale: 1000)
        var actual = CMTime.zero
        guard let cg = try? generator.copyCGImage(at: time, actualTime: &actual) else { return nil }
        return UIImage(cgImage: cg, scale: 1, orientation: .up)
    }
}

final class PoseLocator {
    private let device: String
    private var hist: [Float]?
    private var box: PixelBox?
    private var lastMapped: MappedPeople?
    private var lastTs: Int64 = -1

    init(device: String) {
        self.device = device
    }

    func inferFrame(
        _ image: UIImage,
        tMs: Double,
        seeds: [SeedMark],
        seedBox: NormBox?,
        halfMs: Double,
        markers: PoseLandmarkers
    ) -> PoseFrame {
        let w = PoseImage.pixelWidth(image)
        let h = PoseImage.pixelHeight(image)
        let pixels = PoseImage.argbPixels(image)
        let hit = ClipRange.seedAt(tMs: tMs, seeds: seeds, halfMs: halfMs)
        if let hit {
            box = PersonRoi.denormBox(hit.box, width: w, height: h)
            hist = PersonRoi.buildHist(pixels: pixels, width: w, height: h, box: box!)
        } else if hist == nil, let seedBox {
            box = PersonRoi.denormBox(seedBox, width: w, height: h)
            hist = PersonRoi.buildHist(pixels: pixels, width: w, height: h, box: box!)
        } else if let hist, let box {
            self.box = PersonRoi.search(pixels: pixels, width: w, height: h, hist: hist, prev: box)
        }
        let mapped = detect(image, tMs: Int64(tMs), roi: box, markers: markers)
        if !mapped.poses.isEmpty && !mapped.bboxes.isEmpty {
            let det = mapped.bboxes[0]
            box = PixelBox(x1: Double(det.x1), y1: Double(det.y1), x2: Double(det.x2), y2: Double(det.y2))
            let nextHist = PersonRoi.buildHist(pixels: pixels, width: w, height: h, box: box!)
            hist = hist.map { PersonRoi.blendHist($0, nextHist) } ?? nextHist
            lastMapped = mapped
        } else {
            lastMapped = nil
        }
        let pose = mapped.poses.first
        let blaze = mapped.blazePeople.first ?? []
        return PoseFrame(
            tMs: Int64(tMs),
            poses: pose.map { [$0] } ?? [],
            width: w,
            height: h,
            blaze33: blaze,
            bbox: mapped.bboxes.first
        )
    }

    private func detect(
        _ image: UIImage,
        tMs: Int64,
        roi: PixelBox?,
        markers: PoseLandmarkers
    ) -> MappedPeople {
        let w = PoseImage.pixelWidth(image)
        let h = PoseImage.pixelHeight(image)
        var mapped: MappedPeople?
        if let roi {
            let cropped = detectCrop(image, roi: roi, marker: markers.crop)
            let ok = PoseFilter.isPlausible(cropped, frameW: w, frameH: h) || !cropped.poses.isEmpty
            if ok && !PoseFilter.cropDisagreesWithPrev(lastMapped, cropped) {
                mapped = cropped
            }
        }
        if mapped == nil {
            mapped = detectFull(image, tMs: tMs, marker: markers.video)
            mapped = PoseFilter.pickPrimary(mapped!, prev: lastMapped?.bboxes.first)
            if !PoseFilter.isPlausible(mapped!, frameW: w, frameH: h) {
                let retryBox = mapped!.bboxes.first ?? lastMapped?.bboxes.first
                if let retryBox {
                    let retry = detectCrop(
                        image,
                        roi: PoseFilter.cropBoxForRetry(retryBox, frameW: w, frameH: h),
                        marker: markers.crop
                    )
                    let retryOk = PoseFilter.isPlausible(retry, frameW: w, frameH: h) || !retry.poses.isEmpty
                    if retryOk && !PoseFilter.cropDisagreesWithPrev(lastMapped, retry) {
                        mapped = retry
                    }
                }
                if !PoseFilter.isPlausible(mapped!, frameW: w, frameH: h) && mapped!.poses.isEmpty {
                    mapped = MappedPeople(poses: [], blazePeople: [], bboxes: [])
                }
            }
        }
        return PoseFilter.emaSmooth(lastMapped, mapped!)
    }

    private func detectCrop(_ src: UIImage, roi: PixelBox, marker: any PoseMarker) -> MappedPeople {
        let spec = PersonRoi.cropRect(roi, width: PoseImage.pixelWidth(src), height: PoseImage.pixelHeight(src))
        let srcW = PoseImage.pixelWidth(src)
        let srcH = PoseImage.pixelHeight(src)
        let cw = max(spec.x2 - spec.ox, 2)
        let ch = max(spec.y2 - spec.oy, 2)
        var crop = src
        if !(cw >= srcW - 1 && ch >= srcH - 1) {
            crop = PoseImage.cropped(src, ox: spec.ox, oy: spec.oy, width: cw, height: ch)
        }
        if spec.scale > 1.0 + 1e-6 {
            let sw = max(Int(Double(cw) * spec.scale), 2)
            let sh = max(Int(Double(ch) * spec.scale), 2)
            crop = PoseImage.scaled(crop, width: sw, height: sh)
        }
        let mapped = PoseMapper.fromLandmarks(
            marker.detect(crop),
            width: PoseImage.pixelWidth(crop),
            height: PoseImage.pixelHeight(crop),
            device: device
        )
        let picked = PoseFilter.pickPrimary(mapped, prev: lastMapped?.bboxes.first)
        return remap(picked, spec)
    }

    private func detectFull(_ src: UIImage, tMs: Int64, marker: any PoseMarker) -> MappedPeople {
        let ts = max(tMs, lastTs + 1)
        lastTs = ts
        return PoseMapper.fromLandmarks(
            marker.detectForVideo(src, timestampMs: Int(ts)),
            width: PoseImage.pixelWidth(src),
            height: PoseImage.pixelHeight(src),
            device: device
        )
    }

    private func remap(_ mapped: MappedPeople, _ spec: CropSpec) -> MappedPeople {
        func pt(_ x: Float, _ y: Float) -> (Float, Float) {
            let (nx, ny) = PersonRoi.remapPoint(x: Double(x), y: Double(y), ox: spec.ox, oy: spec.oy, scale: spec.scale)
            return (Float(nx), Float(ny))
        }
        let poses = mapped.poses.map { pose in
            CocoPose(
                keypoints: pose.keypoints.map { kp in
                    let (x, y) = pt(kp.x, kp.y)
                    return CocoKeypoint(x: x, y: y, confidence: kp.confidence)
                }
            )
        }
        let blaze = mapped.blazePeople.map { person in
            person.map { joint in
                let (x, y) = pt(joint.x, joint.y)
                return BlazeJoint(x: x, y: y, z: joint.z, confidence: joint.confidence)
            }
        }
        let boxes = mapped.bboxes.map { box in
            let (x1, y1) = pt(box.x1, box.y1)
            let (x2, y2) = pt(box.x2, box.y2)
            return BBox(x1: x1, y1: y1, x2: x2, y2: y2)
        }
        return MappedPeople(poses: poses, blazePeople: blaze, bboxes: boxes)
    }
}

enum PoseImage {
    static func pixelWidth(_ image: UIImage) -> Int {
        image.cgImage?.width ?? Int(image.size.width * image.scale)
    }

    static func pixelHeight(_ image: UIImage) -> Int {
        image.cgImage?.height ?? Int(image.size.height * image.scale)
    }

    static func argbPixels(_ image: UIImage) -> [UInt32] {
        guard let cg = image.cgImage else { return [] }
        let w = cg.width
        let h = cg.height
        var pixels = [UInt32](repeating: 0, count: w * h)
        let colorSpace = CGColorSpaceCreateDeviceRGB()
        let info = CGBitmapInfo.byteOrder32Little.rawValue | CGImageAlphaInfo.premultipliedFirst.rawValue
        pixels.withUnsafeMutableBytes { raw in
            guard let ctx = CGContext(
                data: raw.baseAddress,
                width: w,
                height: h,
                bitsPerComponent: 8,
                bytesPerRow: w * 4,
                space: colorSpace,
                bitmapInfo: info
            ) else { return }
            ctx.draw(cg, in: CGRect(x: 0, y: 0, width: w, height: h))
        }
        return pixels
    }

    static func cropped(_ image: UIImage, ox: Int, oy: Int, width: Int, height: Int) -> UIImage {
        guard let cg = image.cgImage,
              let part = cg.cropping(to: CGRect(x: ox, y: oy, width: width, height: height))
        else { return image }
        return UIImage(cgImage: part, scale: 1, orientation: .up)
    }

    static func scaled(_ image: UIImage, width: Int, height: Int) -> UIImage {
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = false
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: width, height: height), format: format)
        return renderer.image { _ in
            image.draw(in: CGRect(x: 0, y: 0, width: width, height: height))
        }
    }
}
